// typewriter/build/typewriterbuilder.js
// Generated by CoffeeScript 1.10.0
(function() {
  var Typewriter, TypewriterBuilder, assert;

  assert = require('assert');

  Typewriter = require('./typewriter');

  TypewriterBuilder = function(targetDomElement) {
    var typewriter;
    typewriter = new Typewriter();
    typewriter.setTargetDomElement(targetDomElement);
    return {
      withAccuracy: function(accuracy) {
        this.accuracy = accuracy;
        typewriter.setAccuracy(this.accuracy);
        return this;
      },
      withMinimumSpeed: function(minimumSpeed) {
        this.minimumSpeed = minimumSpeed;
        typewriter.setMinimumSpeed(this.minimumSpeed);
        return this;
      },
      withMaximumSpeed: function(maximumSpeed) {
        this.maximumSpeed = maximumSpeed;
        typewriter.setMaximumSpeed(this.maximumSpeed);
        return this;
      },
      withKeyboardLayout: function(keyboardLayout) {
        this.keyboardLayout = keyboardLayout;
        typewriter.setKeyboardLayout(this.keyboardLayout);
        return this;
      },
      build: function() {
        assert.ok(this.accuracy != null, 'Accuracy must be set');
        assert.ok(this.minimumSpeed != null, 'MinimumSpeed must be set');
        assert.ok(this.maximumSpeed != null, 'MaximumSpeed must be set');
        if (this.keyboardLayout == null) {
          typewriter.setKeyboardLayout(require('./defaultkeyboardlayout'));
        }
        return typewriter;
      }
    };
  };

  module.exports = TypewriterBuilder;

}).call(this);
